### PR TITLE
Feature gg multiline paste

### PR DESCRIPTION
**Pull request by gordonguthrie**

### CODE REVIEW

File: include/riak_shell.hrl [hunk 1]
```diff
@@ -10,11 +10,10 @@
     %% should this command be written to the log and
     %% does this command return an error
     cmd_error              = false :: boolean(),
-    cmd                    = [] :: [char()],
-    log_this_cmd           = true :: boolean(),
-    partial_cmd            = [] :: [char()],
-    partial_tokens         = [] :: [term()],
-    response               = [] :: [char()]
+    cmd                    = []    :: [char()],
```
Comment (andytill) [inline include/riak_shell.hrl:13]: Could be `string()`.

**Comment by andytill** [comment]
As discussed, single quotes are being stripped from queries e.g. `SELECT id FROM docstab WHERE id = 'lol' AND  time > 10 AND time < 1000;`

**Fixed in** 2332430

**Comment by andytill** [comment]
There is a bug when a semi colon appears within single quotes.

```
riak-shell(1)>SELECT id FROM docstab WHERE id = 'l;ol' AND  time > 10 AND time < 1000;
SQL Parser error {<<"ol">>,riak_ql_parser,
                  ["syntax error before: ","identifier"]}
```

Works fine with double quotes.

**UPDATE 14/11/2016**

This is fixed.

``` sql
riak-shell(9)>SELECT a FROM docstab WHERE a = 'l;ol' AND  b > 10 AND b < 1000;        
riak-shell(10)>
```

**Comment by andytill** [comment]
After executing a statement, press enter a few times and then execute another statement. The second statement will fail with `Invalid Command`.

```
riak-shell(1)>CREATE TABLE mtest4( id SINT64 NOT NULL, PRIMARY KEY((id),id));
riak-shell(1)>
riak-shell(1)>
riak-shell(1)>
riak-shell(2)>
riak-shell(2)>
riak-shell(2)>
riak-shell(2)>CREATE TABLE mtest5( id SINT64 NOT NULL, PRIMARY KEY((id),id));
Invalid Command:  CREATE TABLE mtest5( id SINT64 NOT NULL, PRIMARY KEY((id),id));
```

**Fixed in** 7ee5d53.

**Comment by andytill** [comment]
If two commands are on the same line then the second command is ignored.

```
riak-shell(1)>CREATE TABLE mtest8( id SINT64 NOT NULL, PRIMARY KEY((id),id)); SHOW TABLES;
riak-shell(2)>
```

**UPDATE 14/11/2016** create a JIRA for this https://bashoeng.atlassian.net/browse/RTS-1524

**Comment by andytill** [comment]
This change doesn't allow coping in many statements and have them execute one after the other, for example it would be very useful in demo scenarios to run up the shell and paste a page of commands, in this case the descends keys docs I am writing.

``` sql
CREATE TABLE descending_table (
a SINT64 NOT NULL,
b TIMESTAMP NOT NULL,
PRIMARY KEY ((a, quantum(b, 1, 'm')), a, b DESC));

INSERT INTO descending_table VALUES (1,1);
INSERT INTO descending_table VALUES (1,2);
INSERT INTO descending_table VALUES (1,3);
INSERT INTO descending_table VALUES (1,4);
INSERT INTO descending_table VALUES (1,5);

SELECT * FROM descending_table WHERE a = 1 AND b >= 1 AND b <= 5;

+-+------------------------+
|a|           b            |
+-+------------------------+
|1|1970-01-01T00:00:00.005Z|
|1|1970-01-01T00:00:00.004Z|
|1|1970-01-01T00:00:00.003Z|
|1|1970-01-01T00:00:00.002Z|
|1|1970-01-01T00:00:00.001Z|
+-+------------------------+
```

**UPDATE 14/11/2016** create a JIRA for this https://bashoeng.atlassian.net/browse/RTS-1524

**Comment by gordonguthrie** [comment]
`riak-shell` now implements the erlang shell prompt/input now - that doesn't support multi-statement paste - that is an input string with multiple `$end` tokens - because it returns on `$end` so to do this would involve reimplementing a chunk of the io system libraries...

**Comment by gordonguthrie** [comment]
I was going to squash these commits but the merge on the git log makes that a very problematic process.

**Comment by gordonguthrie** [comment]
This would have been the squashed commit message

Major refactor of riak_shell to enable multiline pasting

The initial idea for riak-shell was for a learning exercise for developers moving over to use
leex/yecc - the Erlang lexer/parsers that are used in riak_ql.

In the initial version the prompt returned an input string which could then be lexed.
The lexer tokens were then inspected and, if appropriate, the input string was forwarded
to the riak_ql SQL lexer/parser - if not the riak-shell parser was used.

Switching to the proper Erlang input means that the prompt now returns a token string which must
be reversed to make the input. This made it necessary to turn the cmd_line from a poorly specified
'toy' language to a slightly-less toy one - but one that was better aligned to SQL in how it handled
edge cases.

A lot of the work in this commit is the consequence of 'un-toying' the lexer and all the tech
debt that came with it.

**Comment by andytill** [comment]
Putting a new line in front of a command always results in a command line parser error like below. When repeated the command works fine.

```
riak-shell(2)>INSERT INTO keytab VALUES (2,1); 
riak-shell(3)>INSERT INTO keytab VALUES (3,1);
riak-shell(4)>
riak-shell(4)>INSERT INTO keytab VALUES (4,1);
Error: {error,{2,cmdline_parser,["syntax error before: ",["\"(\""]]}}
```

**Fixed in:** f6ab7dc

**Comment by hazen** [comment]
@andytill @gordonguthrie Any word on this? I'm hoping to tweak `riak_shell`, but need this refactor in before even attempting it. Cheers!

**Comment by andytill** [comment]
@javajolt @gordonguthrie no word, but this fixes a bunch of issues on it's own, so maybe we get this PR merged with the features it has today and JIRA up the other known issues?

**Comment by andytill** [comment]
Verified that this works, but an eunit is failing.

```
riak_shell: concat_atoms_with_underscore_test (module 'riak_shell')...*failed*
in function riak_shell:'-concat_atoms_with_underscore_test/0-fun-0-'/0 (src/riak_shell.erl, line 564)
in call from riak_shell:concat_atoms_with_underscore_test/0 (src/riak_shell.erl, line 563)
**error:{assertMatch_failed,[{module,riak_shell},
                     {line,564},
                     {expression,"make_riak_shell_cmd ( [ { atom , 1 , \"One\" } , { underscore , 1 , \"_\" } , { atom , 1 , \"Two\" } ] )"},
                     {pattern,"{ \"one_two\" , _ }"},
                     {value,{"one",[{underscore,1,"_"},{atom,1,"Two"}]}}]}


=======================================================
  Failed: 1.  Skipped: 0.  Passed: 0.
```

File: src/riak_shell.erl [hunk 2]
```diff
@@ -510,37 +495,63 @@ maybe_print_exception(#state{debug = on}, Type, Err) ->
 maybe_print_exception(_State, _type, _Err) ->
     ok.
 
--ifdef(TEST).
--include_lib("eunit/include/eunit.hrl").
-
-full_cmd_name_test() ->
```
Comment (hazen) [inline src/riak_shell.erl:516]: Do these tests perhaps belong somewhere else?

**Comment by andytill** [comment]
+1

**Comment by hazen** [comment]
+1

**Comment by korry8911** [comment]
+1

**Comment by davidx** [comment]
thumbot retry

**Comment by davidx** [comment]
thumbot retry

**Comment by davidx** [comment]
👍

**Comment by davidpuddy1** [comment]
+1

**Review by davidx — APPROVED** [review]

**Review by davidpuddy1 — APPROVED** [review]

**Comment by davidx** [comment]
+1

**Pull request merged**